File: examples/Experiment06.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.io import mmread
import matplotlib.gridspec as gridspec
from matplotlib.ticker import MaxNLocator

plt.rc('text.latex', preamble=r'\usepackage{amssymb} \usepackage{amsmath}')
plt.rc('text', usetex=True)
plt.rc('font', family='serif')

matrix = "bundle1"        # \in {"bundle1", 
                          #      "4bw100eigs20k", 
                          #      "rand20k"
                          #      "rand20k2"
                          #      "wathen100", 
                          #      "Poisson32k"}
fontsize = 10.5
nreals = 20

res_cg = [np.load("data/Experiment06_" + matrix + "_cg-res_real%d.npz" % (ireal + 1)) for ireal in range(nreals)]
res_pmr = [np.load("data/Experiment06_" + matrix + "_pcg-res_pmr_spai_real%d.npz" % (ireal + 1)) for ireal in range(nreals)]
res_pcg = [np.load("data/Experiment06_" + matrix + "_pcg-res_pcg_spai_real%d.npz" % (ireal + 1)) for ireal in range(nreals)]
res_lopmr = [np.load("data/Experiment06_" + matrix + "_pcg-res_lopmr_spai_real%d.npz" % (ireal + 1)) for ireal in range(nreals)]

#iters_cg = len(res_cg)
#iters_pmr = min(len(res_pmr), 2 * iters_cg)
#iters_pcg = min(len(res_pcg), 2 * iters_cg)
#iters_lopmr = min(len(res_lopmr), 2 * iters_cg)

fig, ax = plt.subplots(figsize=(5.5, 3))

ax.xaxis.set_major_locator(MaxNLocator(integer=True))

#ax.semilogy(np.arange(0, iters_cg), res_cg[:iters_cg], 'r-', label='CG', linewidth=2)
#ax.semilogy(np.arange(0, iters_pmr), res_pmr[:iters_pmr], color='k', linestyle=(0, (3, 1, 1, 1, 1, 1)), label='PCG (PMR-SPAI)', linewidth=2)
#ax.semilogy(np.arange(0, iters_pcg), res_pcg[:iters_pcg], 'k:', label='PCG (PCG-SPAI)', linewidth=2)
#ax.semilogy(np.arange(0, iters_lopmr), res_lopmr[:iters_lopmr], 'k-', label='PCG (LOPMR-SPAI)', linewidth=2)
#ax.semilogy(np.arange(0, iters_pmr), res_pmr[:iters_pmr], color='k', linestyle=(0, (3, 1, 1, 1, 1, 1)), label='PCG (MR)', linewidth=2)
#ax.semilogy(np.arange(0, iters_pcg), res_pcg[:iters_pcg], 'k:', label='PCG (CG)', linewidth=2)
#ax.semilogy(np.arange(0, iters_lopmr), res_lopmr[:iters_lopmr], 'k-', label='PCG (LOMR)', linewidth=2)

lw = .5

ax.semilogy(np.arange(0, len(res_cg[0])), res_cg[0], 'r-', label='pcg (Jacobi)', linewidth=lw)
ax.semilogy(np.arange(0, len(res_pmr[0])), res_pmr[0], 'c-', label='pcg (PMR_SPD SPAI)', linewidth=lw)
ax.semilogy(np.arange(0, len(res_pcg[0])), res_pcg[0], 'k-', label='pcg (PCG SPAI)', linewidth=lw)
ax.semilogy(np.arange(0, len(res_lopmr[0])), res_lopmr[0], 'g-', label='pcg (LOPMR_SPD SPAI)', linewidth=lw)

for ireal in range(1, nreals):
  ax.semilogy(np.arange(0, len(res_cg[ireal])), res_cg[ireal], 'r-', linewidth=lw)
  ax.semilogy(np.arange(0, len(res_pmr[ireal])), res_pmr[ireal], 'c-', linewidth=lw)
  ax.semilogy(np.arange(0, len(res_pcg[ireal])), res_pcg[ireal], 'k-', linewidth=lw)
  ax.semilogy(np.arange(0, len(res_lopmr[ireal])), res_lopmr[ireal], 'g-', linewidth=lw)

ax.set_xlabel('PCG iteration, ' + r'$i$', fontsize=fontsize)
ax.set_ylabel('Backward error, ' + r'$\|r_i\|_2/\|b\|_2$', fontsize=fontsize)
ax.set_title(matrix)
#ax.set_title('Linear solve preconditioned by approximate inverse', fontsize=fontsize)
ax.legend(fontsize=fontsize)
ax.grid(True, alpha=0.3)
ax.tick_params(labelsize=fontsize)

plt.tight_layout()
plt.savefig("img/Experiment06_" + matrix + "_pcg-res_norms.png", bbox_inches='tight', dpi=300)
#plt.savefig("img/Experiment06_" + matrix + "_pcg-res_norms_SIAM_PP26.png", bbox_inches='tight', dpi=300)

File: examples/Experiment06.jl
#push!(LOAD_PATH, ".")
using Pkg
Pkg.activate(".")
using SparseApproximateInverses: pcg

using Random: seed!
using LinearAlgebra
using SparseArrays
using MatrixMarket: mmread, mmwrite
using NPZ


matrix_source = "data/mtcs/spd/"


matrix = "bundle1"        # \in {"bundle1", 
                          #      "4bw100eigs20k", 
                          #      "rand20k"
                          #      "rand20k2"
                          #      "wathen100", 
                          #      "Poisson32k"}


if matrix == "bundle1" # Computer graphics problem
  A = mmread(matrix_source * "bundle1.mtx")
  n = A.n # n = 10,581 | nnz = 770,811
  itmax = 170
elseif matrix == "4bw100eigs20k" # Custom matrix
  A = mmread(matrix_source * "4bw100eigs20k.mtx")
  n = A.n # n = 20,000 | nnz = 179,980
  itmax = 500
elseif matrix == "rand20k"
  A = mmread(matrix_source * "rand20k.mtx")
  n = A.n # n = 20,000 | nnz = 99,772
  itmax = 20
elseif matrix == "rand20k2"
  A = mmread(matrix_source * "rand20k2.mtx")
  n = A.n # n = 20,000 | nnz = 99,772
  itmax = 300
elseif matrix == "wathen100" # Random 2D/3D problem
  A = mmread(matrix_source * "wathen100.mtx")
  n = A.n # n = 30,401 | nnz = 471,601
  itmax = 100
elseif matrix == "Poisson32k" # Random Poisson PDE
  A = mmread(matrix_source * "Poisson32k.mtx")
  n = A.n # n = 31,839 | nnz = 221,375 
  itmax = 200
end


tol = 1e-6
nreals = 20

M = spdiagm(diag(A).^-1)
M_pmr = mmread("data/Experiment05_" * matrix * "_M_pmr_spai_it100.mtx")
M_pmr .+= M_pmr'; M_pmr .*= .5
M_pcg = mmread("data/Experiment05_" * matrix * "_M_pcg_spai_it100.mtx")
M_pcg .+= M_pcg'; M_pcg .*= .5
M_lopmr = mmread("data/Experiment05_" * matrix * "_M_lopmr_spai_it100.mtx")
M_lopmr .+= M_lopmr'; M_lopmr .*= .5


seed!(1)
for ireal in 1:nreals
  b = rand(n)

  x, it, res_norm = pcg(A, b, zeros(n), M, tol, itmax)
  npzwrite("data/Experiment06_" * matrix * "_cg-res_real$ireal.npz", res_norm)

  x, it, res_norm_pmr = pcg(A, b, zeros(n), M_pmr, tol, itmax)
  npzwrite("data/Experiment06_" * matrix * "_pcg-res_pmr_spai_real$ireal.npz", res_norm_pmr)

  x, it, res_norm_pcg = pcg(A, b, zeros(n), M_pcg, tol, itmax)
  npzwrite("data/Experiment06_" * matrix * "_pcg-res_pcg_spai_real$ireal.npz", res_norm_pcg)

  x, it, res_norm_lopmr = pcg(A, b, zeros(n), M_lopmr, tol, itmax)
  npzwrite("data/Experiment06_" * matrix * "_pcg-res_lopmr_spai_real$ireal.npz", res_norm_lopmr)
end